fix: attempt to drain the ReadFileStream for connection pooling

harshavardhana opened the pull request:
## Description
fix: attempt to drain the ReadFileStream for connection pooling

## Motivation and Context
avoid time_wait build up with getObject requests if there are
pending callers and they timeout, can lead to time_wait states

Bonus share the same buffer pool with erasure healing logic,
additionally also fixes a race where parallel readers were
never cleanup during Encode() phase, because pipe.Reader end
was never closed().

Added closer right away upon an error during Encode to make
sure to avoid racy Close() while stream was still being
Read().

## How to test this PR?
go test -v -race shouldn't report any issues

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

poornas commented on cmd/http/close.go:
```diff
@@ -32,12 +32,13 @@ func DrainBody(respBody io.ReadCloser) {
 	// If resp.Body is not closed, the Client's underlying RoundTripper
 	// (typically Transport) may not be able to re-use a persistent TCP
 	// connection to the server for a subsequent "keep-alive" request.
+	maxSplurpSize := int64(2 << 10)
```
```suggestion
	maxSlurpSize := int64(2 << 10)
```

poornas commented on cmd/http/close.go:
```diff
@@ -32,12 +32,13 @@ func DrainBody(respBody io.ReadCloser) {
 		//  - http://stackoverflow.com/a/17961593/4465767
 		defer respBody.Close()
-		io.Copy(ioutil.Discard, respBody)
+		io.CopyN(ioutil.Discard, respBody, maxSplurpSize)
```
```suggestion
		io.CopyN(ioutil.Discard, respBody, maxSlurpSize)
```

harshavardhana committed:
fix: attempt to drain the ReadFileStream for connection pooling

avoid time_wait build up with getObject requests if there are
pending callers and they timeout, can lead to time_wait states

Bonus share the same buffer pool with erasure healing logic,
additionally also fixes a race where parallel readers were
never cleanup during Encode() phase, because pipe.Reader end
was never closed().

Added closer right away upon an error during Encode to make
sure to avoid racy Close() while stream was still being
Read().

klauspost commented on cmd/http/close.go:
```diff
@@ -32,12 +32,13 @@ func DrainBody(respBody io.ReadCloser) {
 		//  - http://stackoverflow.com/a/17961593/4465767
 		defer respBody.Close()
-		io.Copy(ioutil.Discard, respBody)
+		io.CopyN(ioutil.Discard, respBody, maxSlurpSize)
```
Why this limit? AFAICT it will just leave a lot more connections to not be reusable.

It also makes it impossible for it to use the `io.ReaderFrom` interface which is cheaper than the regular copy. 

So I propose to take this out.

harshavardhana replied:
I borrowed it from Go @klauspost as they have used it as a preferable value.

I can take it out as well.

klauspost replied:
@harshavardhana Yeah, let's for now. They are probably conservative for slow-lorris DOS.

klauspost reviewed (approved):
LGTM <3
